drivers: net: native_sim_sockets: new driver

mniestroj opened the pull request:
Add driver for 'native_sim' target that implements offloaded socket
networking by the use of host networking stack and wrapped BSD sockets API.

This driver has following advantages over existing networking drivers for
emulated platforms that are already in tree:
 * no TUN/TAP use means that no additional setup is required on the host
   side:
   * possible to use it within unpriviledged Docker containers, either for
     development or in CI
 * possibility to use and test offloaded sockets
   (CONFIG_NET_SOCKETS_OFFLOAD=y) with emulated target, which allows
   to increase tests coverage of this feature, without requirement of using
   hardware

Native Simulator host libc has different error codes than embedded libc
used by Zephyr. Convert between those.

Depends on (and includes):
- #69347
- #69243
- #69242
- #69237

----

Doc preview:
https://builds.zephyrproject.io/zephyr/pr/65116/docs/boards/posix/native_sim/doc/index.html#nsim-per-offloaded-sockets
https://builds.zephyrproject.io/zephyr/pr/65116/docs/connectivity/networking/native_sim_setup.html#using-offloaded-sockets

mniestroj committed:
drivers: net: nsos: new driver for Native Simulator offloaded sockets

Add driver for 'native_sim' target that implements offloaded socket
networking by the use of host networking stack and wrapped BSD sockets API.

This driver has following advantages over existing networking drivers for
emulated platforms that are already in tree:
 * no TUN/TAP use means that no additional setup is required on the host
   side:
   * possible to use it within unpriviledged Docker containers, either for
     development or in CI
 * possibility to use and test offloaded sockets
   (CONFIG_NET_SOCKETS_OFFLOAD=y) with emulated target, which allows
   to increase tests coverage of this feature, without requirement of using
   hardware

Native Simulator host libc has different error codes than embedded libc
used by Zephyr. Convert between those.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: net: nsos: add DNS offload support

Extend driver to support DNS by offloading getaddrinfo() and freeaddrinfo()
APIs.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
samples: http_get: add tests with Native Simulator offloaded sockets

'http_get' is one of the few samples that just require connection to
internet (google.com over HTTP or HTTPS) and check if HTTP GET was
successful. Use it for minimal coverage of Native Simulator offloaded
sockets:

 * DNS resolving
 * basic TCP transfer
 * compatibility with Zephyr TLS subsystem (which required some fcntl()
   operations)

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

jukkar commented on drivers/net/native_sim.h:
```diff
@@ -0,0 +1,105 @@
+#define NSIM_MSG_DONTWAIT 0x40
+#define NSIM_MSG_WAITALL 0x100
+
+struct nsim_sockaddr {
+	uint16_t sa_family;      /* Address family */
+	char     sa_data[];      /* Socket address */
+};
+
+struct nsim_sockaddr_in {
+	uint16_t sin_family;     /* AF_INET */
+	uint16_t sin_port;       /* Port number */
+	uint32_t sin_addr;       /* IPv4 address */
+};
+
+struct nsim_sockaddr_in6 {
+	uint16_t sin6_family;    /* AF_INET6 */
+	uint16_t sin6_port;      /* Port number */
+	uint32_t sin6_flowinfo;  /* IPv6 flow info */
+	uint8_t  sin6_addr[16];
+	uint32_t sin6_scope_id;  /* Set of interfaces for a scope */
+};
```
Why do we need to have separate nsim sockaddr structs here? I mean sockaddr is a sockaddr always, no?

mniestroj replied:
The problem that the introduction of `struct nsim_*` is trying to solve is to have a common layer between the two libc worlds:
- Native Simulator, which is compiled with host libc
- Zephyr, which is compiled with embedded libc

In order to pass `sockaddr` values between those two worlds, we need a structure that is defined the same way. This is where structs like `struct nsim_sockaddr_in6` are used, since they are locally defined and libc-independent. This is just to provide bridge between two conflicting libc implementations.

aescolar commented on drivers/net/native_sim_errno.c:
```diff
@@ -0,0 +1,123 @@
+/**
```
Sidenote: This is so generic that I'm thinking we should make it common and provided by the native simulator code.
(it can start living here though)

mniestroj replied:
Left it here for now, but it definitely makes sense to move it to native_simulator code whenever there will be a need to reuse it in other context.

jukkar commented on drivers/net/native_sim_errno.h:
```diff
@@ -0,0 +1,95 @@
+
+#include <errno.h>
+
+#define NSIM_EPERM 1         /**< Not owner */
```
Can't we say here
```
#define NSIM_EPERM EPERM
....
```

mniestroj replied:
Similar as described in https://github.com/zephyrproject-rtos/zephyr/pull/65116#discussion_r1391146480, `NSIM_*` values are defined to bridge between Native Simulator and Zephyr libc implementations. If we would define `#define NSIM_EPERM EPERM`, `NSIM_EPERM` value would be potentially different in Zephyr, when there is embedded libc used, and different in Native Simulator when there is host libc used. So by defining `#define NSIM_EPERM 1` we make sure that the value is same in those two contexts.

jukkar replied:
Hmm, don't we need to just convert the errno values between zephyr <-> host, so in either zephyr or in host, meaning we could use the errno value defined in one of them?

aescolar commented on drivers/net/native_sim_adapt.c:
```diff
@@ -0,0 +1,648 @@
+
+void nsim_adapt_freeaddrinfo(struct nsim_addrinfo *nsim_res)
+{
+	struct nsim_addrinfo_wrap *wrap =
```
It feels like the parameter to this function should have been `struct nsim_addrinfo_wrap *wrap` instead of `nsim_res` ? (so you don't need to get its container structure from this pointer)

mniestroj replied:
`struct nsim_addrinfo_wrap` (now `nsos_addrinfo_wrap`) is private to this module, so caller has no knowledge about it.

aescolar commented on drivers/net/native_sim_adapt.c:
```diff
@@ -0,0 +1,648 @@
+		CONTAINER_OF(nsim_res, struct nsim_addrinfo_wrap, nsim_addrinfo);
+
+	freeaddrinfo(wrap->addrinfo);
+	free(wrap);
```
Am i missing something or are both this and the caller free'ing wrap?

mniestroj replied:
There are two separate wrappers, one for Zephyr, one for Linux. I hope that this comment shows what happens there in terms of wrapping:
https://github.com/zephyrproject-rtos/zephyr/blob/b64826ed80683f5b631e2c1b0e27f76670070ce2/drivers/net/nsos_sockets.c#L782-L796

aescolar commented on drivers/net/native_sim_adapt.c:
```diff
@@ -0,0 +1,376 @@
+		*family_host = AF_INET;
+		break;
+	default:
+		return -NSIM_EAFNOSUPPORT;
```
printing an error here would help whoever hits this
Something like:
```
nsi_print_warning("%s: socket family %i not supported\n", __func__, family);
```

Same for the 2 functions below

mniestroj replied:
Added warnings, thanks for suggestion.

aescolar reviewed (changes requested):
A few comments so far

General requests:
* There is no documentation yet in this PR. I guess we should add something in at least these 2:
  * https://docs.zephyrproject.org/latest/connectivity/networking/networking_with_host.html
  * https://docs.zephyrproject.org/latest/boards/posix/native_posix/doc/index.html#peripherals
(note https://github.com/zephyrproject-rtos/zephyr/pull/65057 is moving docs around)

Note: I focus on the "native_simulator" side of things. For the networking side other people are better.

aescolar commented on drivers/net/native_sim_sockets.c:
```diff
@@ -0,0 +1,707 @@
+		*family_nsim = NSIM_AF_INET6;
+		break;
+	default:
+		return -NSIM_EAFNOSUPPORT;
```
Same comment as for _adapt about these error printing something to guide future developers.

mniestroj replied:
I have not added warnings here, since this function is used by `nsos_is_supported()` in order to figure out whether this driver is able to handle specific socket. If not (which is possible for example for TLS sockets), then there would be warning log generated even in a perfectly valid flow.

Also the error codes returned here are very descriptive, i.e. it is clear that either protocol is not supported or socket family is not supported.

aescolar commented on drivers/net/native_sim_adapt.c:
```diff
@@ -0,0 +1,648 @@
+		host_fds[i].revents = fds[i].revents;
+	}
+
+	ret = poll(host_fds, nfds, timeout);
```
This is not going to work as you expect if there is any timeout. (You are going to hang the whole zephyr until the call to linux's poll() returns)

I presume if there is a timeout and it is not ready, you'd want it to k_yield().
So, this would need to be replaced with a while loop (in the caller), where this calls the linux poll() with a timeout of 0, and (in the caller) while there is time left and the descriptor is not ready (ret==0), you k_sleep() for some kconfigured amount of time, before calling this function to do a the linux poll() again.

mniestroj replied:
You were right. i have implemented your suggestion with `k_sleep()` on the Zephyr (caller) in the first place. However, then I reworked the code to utilize `NSI_HW_EVENT()`, which seems the perfect mechanism for polling host interfaces. The advantage of the latter is that there is minimal latency between traffic on the host and notification of Zephyr. This makes it much more performant (better network transfer speeds). Additionally it makes it independent of systick/counter/timer configuration, which affects `k_sleep()` behavior and its latency.

aescolar reviewed (commented):
Some extra comments

jukkar commented on drivers/net/native_sim_adapt.c:
```diff
@@ -0,0 +1,376 @@
+/**
```
I have used "adapt" (as adaptation) in other files that interface with Linux in various networking drivers. So this wording would be ok for me and in line with other networking drivers.

mniestroj replied:
I have left file names as they were, but added comment at the top of the file describing more precisely what each file implements.

jukkar commented on drivers/net/native_sim_sockets.c:
```diff
@@ -0,0 +1,533 @@
+{
+	switch (family) {
+	case AF_UNSPEC:
+		*family_nsim = NSIM_AF_UNSPEC;
```
I am getting all the time confused with the native word in various places. It does not give much context, at least for me. It would be great if it is obvious where the code is to be used (zephyr vs host side).

mniestroj replied:
I have changed all those "middleground" macros and conversion functions to use `NSOS_MID_` and `nsos_mid_`. I hope that improves readability.

beriberikix commented:
Please remove the `stale` label. Thanks!

mniestroj committed:
drivers: net: nsos: implement poll() syscall

Use NSI_HW_EVENT() in order to periodically check for events in host
sockets. Whenever there is a socket event ready to be processed by Zephyr,
raise native_sim (newly introduced) CPU interrupt, so that Zephyr driver
can signal readiness with k_poll().

Maintain a list of Zephyr poll() executions in Zephyr context. Iterate
through them whenever there is some event to be processed.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: net: nsos: add F_GETFL and F_SETFL support

This allows to use Zephyr TLS subsystem with Native Sim offloaded sockets.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: net: nsos: add IPv6 support

Add support for IPv6 socket offloading, next to existing IPv4 support.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: net: nsos: support blocking accept() and recvfrom()

So far only non-blocking accept() and recvfrom() were suported. This patch
implements blocking behavior, with the use of poll(fd, POLLIN) as helper
mechanism.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
samples: net: echo_server: add config for native_sim offloaded sockets

Add configuration for Native Simulator offloaded sockets.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
samples: net: echo_server: add native_sim to allowed platforms

'native_sim' builds without any issues as part of twister and it also works
when manually setting up Ethernet virtual device (using
tools/net-tools/net-setup.sh).

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
doc: networking: document native_sim with offloaded sockets

Document how offloaded sockets can be used with 'native_sim' for networking
capability.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
boards: native_sim: document offloaded sockets driver

Document offloaded sockets driver, which is an alternative to Ethernet
driver for networking applications.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>
Co-authored-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

mniestroj commented on drivers/net/Kconfig:
```diff
@@ -228,4 +228,14 @@ config NET_CANBUS_INIT_PRIORITY
+# Native simulator offloaded sockets
+#
+
+config NET_NATIVE_SIM_OFFLOADED_SOCKETS
```
Used that name, as you suggested.

mniestroj commented on drivers/net/Kconfig:
```diff
@@ -228,4 +228,14 @@ config NET_CANBUS_INIT_PRIORITY
+#
+
+config NET_NATIVE_SIM_OFFLOADED_SOCKETS
+	bool "Native Simulator offloaded sockets"
```
Added help message.

mniestroj commented on drivers/net/native_sim.h:
```diff
@@ -0,0 +1,105 @@
+/**
```
I have renamed all files and changed the namespace to "nsos", as you suggested in https://github.com/zephyrproject-rtos/zephyr/pull/65116#discussion_r1391065207

mniestroj commented on drivers/net/native_sim_sockets.c:
```diff
@@ -0,0 +1,533 @@
+{
+	int ret;
+
+	ret = nsim_adapt_read(OBJ_TO_FD(obj), buf, sz);
```
Applied that idea and it reduces implementation size and number of functions in general. Thanks for suggestion!

aescolar commented on drivers/net/nsos_sockets.c:
```diff
@@ -0,0 +1,947 @@
+ * Zephyr (top) side of NSOS (Native Simulator Offloaded Sockets).
+ */
+
+#define _POSIX_C_SOURCE 200809
```
Better undef it before, just to avoid a possible redefinition warning if somebody happens to define it globally.
(undefining is safe even if it was never set)

mniestroj replied:
Added `#undef _POSIX_C_SOURCE` and changed value to `200809L`.

aescolar reviewed (changes requested):
(Just a partial review so far)
Thanks @mniestroj we are getting closer.
Note the comment about the zephyr include in the bottom, and the one about polling every µs.

jukkar commented on drivers/net/nsos_sockets.c:
```diff
@@ -0,0 +1,525 @@
+	return ret;
+}
+
+static ssize_t nsos_sendmsg(void *obj, const struct msghdr *msg, int flags)
```
We currently also support `recvmsg` so please add `nsos_recvmsg` and return `ENOTSUP` for it too

mniestroj replied:
Added `nsos_recvmsg`, as suggested.

jukkar commented on boards/posix/native_sim/doc/index.rst:
```diff
@@ -379,6 +379,13 @@ The following peripherals are currently provided with this board:
    https://github.com/zephyrproject-rtos/net-tools
 
+**Offloaded sockets driver**
+  Offloaded sockets driver is an alternative to Ethernet driver. Instead of
```
Yeah, perhaps the suggestion is nicer, although I can fully understand the original text too.

kartben replied:
Maybe more important would be to document/explain the pros and cons of each approach?

What would one be missing (if anything) when using this new approach over TUN? If there's hardly any limitation or "cons", maybe it would make sense to actually *recommend* people favour this new driver as it's much easier to setup?

kartben replied:
> One of the "cons" for this driver is that we avoid using the L2 layer, Ethernet in this case. This means that various parts of the stack are not tested / utilized, also using VLAN, ARP, gPTP, LLDP etc is not possible with this. So if only testing / using application protocols and not caring about lower level stuff, then this driver is nice. Big benefit is also that user does not need root access, so this makes running tests etc. in container much nicer.

exactly the kind of answer I was hoping to get... and that I would like to see captured in the docs somewhere! :) FWIW it can totally be a follow-up PR. It might work quite well to capture the pros/cons in table form, with something like ❌ / ✅ for the various items/features

mniestroj replied:
Applied @aescolar suggestion.

aescolar replied:
Thanks.
For @kartben 's request, would it make sense to add something like:
``` 
  This driver is an alternative to the :ref:`TAP based ethernet driver
  <nsim_per_ethe>`. Instead of using a virtual network in the Linux side, this
  driver utilizes Linux's standard BSD socket API. With this, multiple Zephyr
  applications can communicate over the Linux loopback interface.
  The benefit of this approach is that root privileges are not required and
  that the process is connected to the same interface as other Linux processes
  instead of a virtual network, facilitating testing without the need for extra
  setup in the host. The drawback is that the L2 layer of Zephyr's networking
  stack is not exercised.
 ``` 
 ( just a proposal, either of you can have better and more accurate wording :) )

mniestroj replied:
Extended description with content above.

jukkar reviewed (commented):
Minor nit added. Impressive work, LGTM!

aescolar commented on samples/net/sockets/echo_server/overlay-nsos.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_ETH_NATIVE_POSIX=n
```
(sidetrack) (@jukkar ) I wonder if after we have this new driver, if it we should change defaults. What do you think?

jukkar replied:
> I wonder if after we have this new driver, if it we should change defaults. What do you think?

@aescolar What defaults do you mean, could you elaborate a bit more?

aescolar replied:
I guess that I was thinking that instead of this default https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/posix/native_sim/Kconfig.defconfig#L22
we'd default to this driver `if NETWORKING`. (But you know much more about this kconfigs and what is better to test more or offer by default, so it was just an open question)

jukkar replied:
> I guess that I was thinking that instead of this default https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/posix/native_sim/Kconfig.defconfig#L22

That is certainly possible, but that needs to be discussed separately after we have merged this and seen how to use it. After all, this change is somewhat specialized as it bypasses L2 so it depends on use cases when to use it or not.

aescolar commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+
+	ret = fcntl(fd, F_SETFL, fl_from_nsos_mid(flags));
+	if (ret < 0) {
+		return -errno;
```
does this errno need conversion?

aescolar commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+		sockaddr_to_nsos_mid(res_p->ai_addr, res_p->ai_addrlen,
+				     wrap->addrinfo_mid.ai_addr, &wrap->addrinfo_mid.ai_addrlen);
+
+		wrap->addrinfo_mid.ai_canonname =
```
maybe a stupid question, but: when will this be free'd?

mniestroj replied:
Not a stupid queston at all. This was indeed a memory leak. Added:
```
for (struct nsos_mid_addrinfo *res_p = res_mid; res_p; res_p = res_p->ai_next) {
	free(res_p->ai_canonname);
}
```
in `nsos_adapt_freeaddrinfo()`. Updated `nsos_sockets.c` as well.

aescolar commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+	}
+	}
+
+	return -NSOS_MID_EINVAL;
```
seems ignored in all calls(?)

mniestroj replied:
Updated all callers to take this return value into account.

aescolar commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+static int sockaddr_to_nsos_mid(const struct sockaddr *addr, socklen_t addrlen,
+				struct nsos_mid_sockaddr *addr_mid, size_t *addrlen_mid)
+{
+	if (!addr) {
```
Would a check for `addrlen == 0` be necessary?
(similar comment in `sockaddr_from_nsos_mid()`)
Running the new sample under valgrind gives:
```
==3113887==    at 0x80689CB: sockaddr_to_nsos_mid (nsos_adapt.c:303)
==3113887==    by 0x8068E8C: nsos_adapt_recvfrom (nsos_adapt.c:451)
==3113887==    by 0x80617FB: nsos_recvfrom_with_poll (nsos_sockets.c:691)
==3113887==    by 0x80617FB: nsos_recvfrom (nsos_sockets.c:717)
==3113887==    by 0x804E14F: z_impl_zsock_recvfrom (sockets.c:1773)
==3113887==    by 0x8049756: zsock_recvfrom (socket.h:313)
==3113887==    by 0x8049756: zsock_recv (socket.h:547)
==3113887==    by 0x8049756: recv (socket.h:890)
==3113887==    by 0x8049756: main (http_get.c:115)
```

mniestroj replied:
Indeed, that was necessary. Stumbled upon that after checking error codes from `sockaddr_to_nsos_mid()`. So... added check for `addrlen == 0`, as suggested.

aescolar commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+		nsi_print_warning("error in nsos_adapt poll(): errno=%d\n", errno);
+
+		nsos_adapt_poll_time = NSI_NEVER;
+		nsi_hws_find_next_event();
```
nit: this call is not needed (neither in 639)

mniestroj replied:
Removed both invocations.

aescolar commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+
+	ret = poll(fds, nfds, 0);
+	if (ret < 0) {
+		nsi_print_warning("error in nsos_adapt poll(): errno=%d\n", errno);
```
Question: should this be an error? (given that you disable the functionality after this)

mniestroj replied:
Extended poll() error handling with two cases:
- `EINTR` seems to be "recoverable", so left a warning message and changed `nsos_adapt_poll_time` to be `nsi_hws_get_time() + 1`
- other error codes: calling `nsi_print_error_and_exit()` instead, since further application run is most likely not reliable/useful

aescolar reviewed (dismissed):
Thanks again @mniestroj 
A few extra comments

aescolar commented on scripts/native_simulator/native/src/include/nsi_cpu0_interrupts.h:
```diff
@@ -9,6 +9,7 @@
 #define TIMER_TICK_IRQ 0
 #define OFFLOAD_SW_IRQ 1
 #define COUNTER_EVENT_IRQ 2
+#define NSOS_IRQ 3
```
Better define it in https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/posix/native_sim/board_soc.h#L24 
and include soc.h in nsos_adapt.c
(This nsos_adapt code is zephyr specific, while this file is part of the upstream native simulator. Booking interrupts upstream for Zephyr only use is a bit out of place)

mniestroj replied:
Moved this definition to `board_soc.h` as suggested. Should it still be value of 3 or is that range reserved for upstream interrupts only?

aescolar replied:
By now we can just let it be a hardcoded 3. Later I can add a macro in the native simulator header keeping the first available interrupt number for the user, and then we can define it based on that.

mniestroj commented on doc/connectivity/networking/native_sim_setup.rst:
```diff
@@ -89,3 +92,55 @@ You can manually connect to it like this:
+Using offloaded sockets
+***********************
+
+Main advantage over `Using virtual Ethernet driver`_ is no need to setup
```
Applied your suggestion.

mniestroj commented on doc/connectivity/networking/native_sim_setup.rst:
```diff
@@ -89,3 +92,55 @@ You can manually connect to it like this:
+
+.. note::
+
+   Native Simulator with offloaded sockets is using the same network
```
Applied your suggestion.

mniestroj commented on doc/connectivity/networking/networking_with_host.rst:
```diff
@@ -59,7 +59,8 @@ possible:
     can attach host debugger directly to the running Zephyr instance. This
     requires that there is an adaptation driver in Zephyr for interfacing
-    with the host system. An Ethernet driver exists in Zephyr for this purpose.
+    with the host system. An Ethernet and offloaded sockets drivers exist in
```
Applied your suggestion.

mniestroj commented on drivers/net/CMakeLists.txt:
```diff
@@ -9,3 +9,20 @@ if(CONFIG_NET_NATIVE)
+  zephyr_library_sources(nsos_errno.c)
+  zephyr_library_sources(nsos_fcntl.c)
+  zephyr_library_sources(nsos_sockets.c)
+  target_compile_options(native_simulator BEFORE INTERFACE
```
Dropped the use of sys_dlist, as it is no longer needed with switch from 'poll' to 'epoll' on Linux side.

mniestroj commented on drivers/net/Kconfig:
```diff
@@ -225,4 +225,20 @@ config NET_CANBUS_INIT_PRIORITY
+	  Offloaded sockets for Native Simulator utilize host BSD sockets API (like socket(),
+	  connect(), send(), recvfrom(), etc.) in order to provide networking capability.
+
+	  Usage of this driver (as compared to others) is possible without any additional setup on
```
Applied your suggestion.

mniestroj commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+		hw_irq_ctrl_set_irq(NSOS_IRQ);
+	}
+
+	nsos_adapt_poll_time = nsi_hws_get_time() + 1;
```
Added a Kconfig option with default of 1000 cycles after no events were detected in previous (e)poll execution.

mniestroj commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+	int nfds = sys_dlist_len(&nsos_adapt_sockets);
+	int ret;
+
+	fds = realloc(nsos_adapt_fds, sizeof(*fds) * nfds);
```
I have reworked the implementation to use 'epoll' instead of 'poll'. That way we don't need to track monitored file descriptors (kernel does that for us) and that made it possible to drop the use of linked list as well as realloc call here.

mniestroj commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+
+#include <zephyr/sys/dlist.h>
+
+static sys_dlist_t nsos_adapt_sockets = SYS_DLIST_STATIC_INIT(&nsos_adapt_sockets);
```
I have reworked the implementation to use 'epoll' instead of 'poll'. That way we don't need to track monitored file descriptors (kernel does that for us) and that made it possible to drop use of linked list.

mniestroj commented on drivers/net/nsos_adapt.c:
```diff
@@ -0,0 +1,673 @@
+	static struct pollfd *nsos_adapt_fds;
+	struct nsos_mid_pollfd *pollfd_mid;
+	struct pollfd *fds;
+	int nfds = sys_dlist_len(&nsos_adapt_sockets);
```
Added `nsos_adapt_nfds` to store number of monitored file descriptors.

mniestroj commented on drivers/net/nsos.h:
```diff
@@ -0,0 +1,135 @@
+
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/types.h>
```
Removed this include statement.

mniestroj commented on drivers/net/nsos_fcntl.c:
```diff
@@ -0,0 +1,72 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Added description to both `nsos_fcntl.c` and `nsos_netdb.c`. Open to any feedback on that.

mniestroj commented on drivers/net/nsos_dns.c:
```diff
@@ -0,0 +1,65 @@
+/**
```
Changed file names to `nsos_netdb.{c,h}`, as suggested.

mniestroj commented on boards/posix/native_sim/doc/index.rst:
```diff
@@ -379,6 +379,13 @@ The following peripherals are currently provided with this board:
+  offloaded sockets driver utilizes network access just as usual Linux
+  application, using BSD sockets API. Multiple Zephyr applications can
+  communicate that way over loopback interface.
+
```
Added following line:
```
Offloaded sockets, :ref:`nsim_per_offloaded_sockets`, :kconfig:option:`CONFIG_NET_NATIVE_OFFLOADED_SOCKETS`, All
```

mniestroj commented on drivers/net/nsos_sockets.c:
```diff
@@ -0,0 +1,947 @@
+		return 0;
+	}
+
+	res_wraps = calloc(n_res, sizeof(*res_wraps));
```
This was meant to use embedded libc on Zephyr side. However since I used `k_malloc()` in other place, it makes sense to use `k_calloc()` here as well. Also replaced `free()` with `k_free()`.

mniestroj commented:
Thanks @aescolar and @jukkar for thorough review. I tried to address all your suggestions so far.

FYI in the next 2 weeks I will not be available, so the next round of feedback I will likely address in ~3 weeks. Thanks for patience.

aescolar reviewed (commented):
Thanks @mniestroj 
Just a few quite minor things left

mniestroj commented on doc/connectivity/networking/networking_with_host.rst:
```diff
@@ -59,7 +59,8 @@ possible:
     requires that there is an adaptation driver in Zephyr for interfacing
-    with the host system. An Ethernet driver exists in Zephyr for this purpose.
+    with the host system. Two possible network drivers can be used for this
+    purpose, a TAP virtual Ethernet driver and an offloaded sockets driver."
```
Removed, thank for spotting that!

mniestroj commented on drivers/net/Kconfig:
```diff
@@ -225,4 +225,34 @@ config NET_CANBUS_INIT_PRIORITY
+if NET_NATIVE_OFFLOADED_SOCKETS
+
+config NET_NATIVE_OFFLOADED_SOCKETS_EPOLL_WAIT_INTERVAL
+	int "Interval between epoll_wait() calls (in cycles)"
```
Applied.

mniestroj commented on drivers/net/Kconfig:
```diff
@@ -225,4 +225,34 @@ config NET_CANBUS_INIT_PRIORITY
+	int "Interval between epoll_wait() calls (in cycles)"
+	default 1000
+	help
+	  Number of Native Simulator cycles before next epoll_wait() call, when there were no
```
Applied.

aescolar commented:
The compliance check is a false positive which will need waiving.

----

@jukkar @rlubos To me it looks good, your review and blessing as networking maintainers would be the missing step.

nordicjm reviewed (changes requested):
Needs converting to hwmv2

aescolar commented:
Please rebase this PR.
Now that the hwmv2 branch has been merged this PR needs rebasing on top of the latest main.
Note that the boards & socs folder has moved, and many boards names as used in tests yaml
and overlays have changed.

aescolar commented:
(The checkpatch warning is a false positive which should be waived)

@nordicjm the PR has been rebased, please re-review or remove the rejection
@rlubos @jukkar your review as maintainers of the networking subsystem would be very welcome

nordicjm reviewed (approved):
Build system parts OK
